aten/src/ATen/core/type.cpp

```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+  // possible types
+  std::sort(types.begin(), types.end(),
+          [](const TypePtr a, const TypePtr b) -> bool {
+            if (a->kind() != b->kind()) {
+              return a->kind() < b->kind();
+            }
+            return a->str() < b->str();
+          });
+}
+
+UnionType::UnionType(std::vector<TypePtr> types, TypeKind kind) : Type(kind) {
+  for (auto type : types) {
+    flattenUnion(type, types_);
+  }
+
+  standardizeUnion(types_);
+
+  TORCH_INTERNAL_ASSERT(types_.size() != 1, "After type unification was"
+                        " performed, the Union only has one type. "
+                        "Use the common supertype instead of creating "
+                        "a Union type");
+
+  TORCH_INTERNAL_ASSERT(types_.size() >= 2, "Cannot create a Union of "
+                        "one or fewer types");
+
+  can_hold_none_ = false;
+  has_free_variables_ = false;
+
+  for (const TypePtr type : types_) {
+    if (type->kind() == NoneType::Kind) {
+      can_hold_none_ = true;
+    }
+    if (type->hasFreeVariables()) {
+      has_free_variables_ = true;
+    }
+  }
+
+}
+
+UnionTypePtr UnionType::create(std::vector<TypePtr> types) {
+  auto union_type = new UnionType(std::move(types));
+  if (auto optional_type = union_type->to_optional()) {
+    return OptionalType::create(std::move(*optional_type));
+  }
+  return UnionTypePtr(std::move(union_type));
+}
+
+bool UnionType::operator==(const Type& rhs) const {
+  if (auto union_rhs = rhs.cast<UnionType>()) {
+    return types_ == union_rhs->types_;
+  }
+  if (auto optional_rhs = rhs.cast<OptionalType>()) {
+    auto optional_lhs = this->to_optional();
+    if (optional_lhs) {
+      return *optional_rhs == *((optional_lhs.value())->cast<OptionalType>());
+    }
+    return false;
+  }
+  return false;
+}
+
+bool UnionType::isSubtypeOfExt(const TypePtr& rhs_, std::ostream* why_not) const {
+  if (const auto union_rhs = rhs_->cast<UnionType>()) {
+    // Fast path: The Union `types` vector is always sorted, so we can
+    // do an O(n) linear comparison of both vectors. We need to use
+    // `std::equal` with `TypeEqual` for optimal performance; do not
+    // change this to `operator==`!
+    if (std::equal(this->types_.begin(), this->types_.end(),
+                      union_rhs->types().begin(), union_rhs->types().end(),
+                      TypeEqual())) {
+      return true;
+    }
+    // Final O(n^2) check since the vector equality comparison above
+    // won't handle subtyping relationships
+    return std::all_of(this->types_.begin(), this->types_.end(),
+      [&](TypePtr this_type) -> bool {
+        return this_type->isSubtypeOf(union_rhs);
+    });
+  } else if (const auto optional_rhs = rhs_->cast<OptionalType>()) {
+    TypePtr inner = optional_rhs->getElementType();
+    return can_hold_none_ && types_.size() == 2 &&
+      (types_[0]->isSubtypeOf(inner) ^ types_[1]->isSubtypeOf(inner));
```

Comment:
Can `Optional` can't have an `inner` type that None subtypes? I think that would be a malformed type. Still i think this could be more clear to follow with `(types_[0] != Nonetype::get() && types_[0]->isSubtypeOf(inner)) || (types_[1] != Nonetype::get() && types_[1]->isSubtypeOf(inner))`